include/seastar/core/shared_mutex.hh

```diff
@@ -234,6 +234,122 @@ with_lock(shared_mutex& sm, Func&& func) noexcept {
     }
 }
 
+/// \brief RAII-based shared lock for a \ref shared_mutex object.
+///
+/// As soon as the destructor of an object of this class is called,
+/// the corresponding \ref shared_mutex is shared unlocked.
+///
+/// The class is effectively a safe way to keep a \ref shared_mutex
+/// locked and to unconditionally release it because no matter what
+/// happens, the destructor should eventually be called.
+///
+/// The corresponding \ref shared_mutex object MUST be kept alive
+/// throughout the lifetime of this object.
+///
+/// The class is NOT responsible for shared locking a \ref shared_mutex.
+/// \see get_shared_lock(shared_mutex&)
+/// \see unique_lock
+class shared_lock {
+private:
+    friend future<shared_lock> get_shared_lock(shared_mutex& mutex);
+private:
+    shared_mutex* _mutex;
+
+private:
+    shared_lock(shared_mutex& mutex) noexcept : _mutex{std::addressof(mutex)} {}
+public:
+    shared_lock(shared_lock&& other) noexcept : _mutex{std::exchange(other._mutex, nullptr)} {}
+    shared_lock& operator=(shared_lock&& other) noexcept {
+        if (this != std::addressof(other)) {
+            this->~shared_lock();
+            new (this) shared_lock{std::move(other)};
+        }
+        return *this;
+    }
+
+    shared_lock(const shared_lock&) = delete;
+    shared_lock& operator=(const shared_lock&) = delete;
+
+    ~shared_lock() noexcept {
+        if (_mutex) {
+            _mutex->unlock_shared();
+            _mutex = nullptr;
```

Comment:
not really needed in the destructor,
better keep it as is so that the sanitizer can detect use after free if somehow double destruction happens.